src/frontend/packages/activitypub-components/src/utils/shaclValidation.ts

```diff
@@ -0,0 +1,132 @@
+const stringToStream = (str: string): ReadableWebToNodeStream => {
+  // Create a TextEncoder to convert string to Uint8Array
+  const encoder: TextEncoder = new TextEncoder();
+  const uint8Array: Uint8Array = encoder.encode(str);
+
+  // Create a ReadableStream from the Uint8Array
+  const readableStream: ReadableStream = new ReadableStream({
+    start(controller: ReadableStreamDefaultController) {
+      controller.enqueue(uint8Array);
+      controller.close();
+    }
+  });
+
+  // Convert the web ReadableStream to a Node.js Readable stream
+  return new ReadableWebToNodeStream(readableStream);
+};
+
+// Helper function to parse JSON-LD and convert to RDF-ext quads
+const parseJsonLd = async (jsonLdObj: object, context: string[]) => {
+  try {
+    // Add context to the JSON-LD object if needed
+    const jsonLdWithContext = {
+      ...jsonLdObj,
+      '@context': context
+    };
+
+    // Convert JSON-LD object to string
+    const jsonString: string = JSON.stringify(jsonLdWithContext);
+
+    // Convert string to stream using the helper function
+    const textStream: ReadableWebToNodeStream = stringToStream(jsonString);
+
+    // Use the JsonLdParser that outputs rdf-ext compatible quads directly
+    const parser: JsonLdParser = new JsonLdParser({ factory: rdf });
+    const quadStream: any = parser.import(textStream);
+
+    // Collect quads into a dataset
+    const dataset: any = rdf.dataset();
+    for await (const quad of quadStream) {
+      dataset.add(quad);
+    }
+
+    return dataset;
+  } catch (error) {
+    console.error('Error parsing JSON-LD:', error);
+    throw error;
+  }
+};
+
+// Helper function to load a SHACL shape and return a validator
+const getShaclValidator = async (shapeUri: string): Promise<SHACLValidator> => {
+  // Check if the validator is already cached
+  if (validatorCache[shapeUri]) {
+    return validatorCache[shapeUri];
+  }
+
+  const response = await fetch(shapeUri, {
+    headers: {
+      Accept: 'text/turtle'
+    }
+  });
+  if (!response.ok) {
+    throw new Error(`Failed to load shape: ${response.status} ${response.statusText}`);
+  }
+
+  // Get the Turtle data as text
+  const turtleData: string = await response.text();
+
+  // Convert text to stream using the helper function
+  const textStream: ReadableWebToNodeStream = stringToStream(turtleData);
+
+  // Use ParserN3 which outputs rdf-ext compatible quads directly
+  const parser: ParserN3 = new ParserN3({ factory: rdf });
+  const quadStream: any = parser.import(textStream);
+
+  // Collect quads into a dataset
+  const shapeDataset = rdf.dataset();
+  for await (const quad of quadStream) {
+    shapeDataset.add(quad);
+  }
+
+  // Create and cache the SHACL validator using the dataset
+  validatorCache[shapeUri] = new SHACLValidator(shapeDataset, { factory: rdf });
+  return validatorCache[shapeUri];
+};
+
+const validateItems = async (
+  items: Array<object>,
+  shaclValidator: SHACLValidator,
+  context: string[]
```

Comment:
A context can be more complex. On the semantic-data-provider, we use this: `string | string[] | Record<string, string>`